langstream-runtime/langstream-runtime-impl/src/main/java/ai/langstream/runtime/agent/GrpcAgentProcessor.java

```diff
@@ -0,0 +1,386 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package ai.langstream.runtime.agent;
+
+import ai.langstream.api.runner.code.AbstractAgentCode;
+import ai.langstream.api.runner.code.AgentProcessor;
+import ai.langstream.api.runner.code.Header;
+import ai.langstream.api.runner.code.Record;
+import ai.langstream.api.runner.code.RecordSink;
+import ai.langstream.api.runner.code.SimpleRecord;
+import ai.langstream.runtime.grpc.ProcessorGrpc;
+import ai.langstream.runtime.grpc.ProcessorGrpc.ProcessorStub;
+import ai.langstream.runtime.grpc.ProcessorOuterClass;
+import ai.langstream.runtime.grpc.ProcessorOuterClass.ProcessorRequest;
+import ai.langstream.runtime.grpc.ProcessorOuterClass.ProcessorResponse;
+import ai.langstream.runtime.grpc.ProcessorOuterClass.ProcessorResult;
+import ai.langstream.runtime.grpc.ProcessorOuterClass.Records;
+import ai.langstream.runtime.grpc.ProcessorOuterClass.Value;
+import com.fasterxml.jackson.databind.JsonNode;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.protobuf.ByteString;
+import io.grpc.ManagedChannel;
+import io.grpc.ManagedChannelBuilder;
+import io.grpc.stub.StreamObserver;
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.atomic.AtomicInteger;
+import lombok.extern.slf4j.Slf4j;
+import org.apache.avro.Conversions;
+import org.apache.avro.Schema;
+import org.apache.avro.generic.GenericDatumReader;
+import org.apache.avro.generic.GenericDatumWriter;
+import org.apache.avro.generic.GenericRecord;
+import org.apache.avro.io.BinaryEncoder;
+import org.apache.avro.io.Decoder;
+import org.apache.avro.io.DecoderFactory;
+import org.apache.avro.io.EncoderFactory;
+
+@Slf4j
+public class GrpcAgentProcessor extends AbstractAgentCode implements AgentProcessor {
+
+    private ManagedChannel channel;
+    private ProcessorStub processorStub;
+    private StreamObserver<ProcessorRequest> request;
+    private RecordSink sink;
+
+    // For each record sent, we increment the recordId
+    private final AtomicInteger recordId = new AtomicInteger(0);
+
+    // For each record sent, we store the record and the sink to which the result should be emitted
+    private final Map<Long, RecordAndSink> sourceRecords = new ConcurrentHashMap<>();
+
+    // For each schema sent, we increment the schemaId
+    private final AtomicInteger schemaId = new AtomicInteger(0);
+
+    // Schemas sent to the server
+    private final Map<Object, Integer> schemaIds = new ConcurrentHashMap<>();
+
+    // Schemas received from the server
+    private final Map<Integer, Object> serverSchemas = new ConcurrentHashMap<>();
```

Comment:
An untrusted remote server could OOM us by sending a big amount of schemas.
I don't think this is an issue atm since the remote application is owned by the user.
But in the future we may have to put a limit.